libs/zedUpload/httputil/http.go

```diff
@@ -346,3 +195,149 @@ func ExecCmd(ctx context.Context, cmd, host, remoteFile, localFile string, objSi
+		}
+		// we need innerCtx cancel to call in case of inactivity
+		innerCtx, innerCtxCancel := context.WithCancel(ctx)
+		inactivityTimer := time.AfterFunc(inactivityTimeout, func() {
+			//keep it to call cancel regardless of logic to releases resources
+			innerCtxCancel()
+		})
+		req, err := http.NewRequestWithContext(innerCtx, http.MethodGet, host, nil)
+		if err != nil {
+			stats.Error = fmt.Errorf("request failed for get %s: %s",
+				host, err)
+			return stats, Resp{}
+		}
+		req.Header.Set("User-Agent", userAgent)
+		req.Header.Set("Content-Type", "application/octet-stream")
+
+		withRange := false
+		//add Range header if server supports it and we already receive data
+		if supportRange && copiedSize > 0 {
+			withRange = true
+			req.Header.Set("Range", fmt.Sprintf("bytes=%d-", copiedSize))
+		}
+		resp, err := client.Do(req)
+		if err != nil {
+			// break the retries loop and skip the error from
+			// http *net.DNSError if the error has the suffix
+			// of "no suitable address found"
+			if IsNoSuitableAddrErr(err) {
+				appendToErrorList(attempt, fmt.Errorf(NoSuitableAddrStr))
+				break
+			}
+			appendToErrorList(attempt, fmt.Errorf("client.Do failed: %s", err))
+			continue
+		}
+
+		// supportRange indicates if server supports range requests
+		supportRange = resp.Header.Get("Accept-Ranges") == "bytes"
+
+		//if we not receive StatusOK for request without Range header or StatusPartialContent for request with range
+		//it indicates that server misconfigured
+		if !withRange && resp.StatusCode != http.StatusOK || withRange && resp.StatusCode != http.StatusPartialContent {
+			respErr := fmt.Errorf("bad response code: %d", resp.StatusCode)
+			err = resp.Body.Close()
+			if err != nil {
+				respErr = fmt.Errorf("respErr: %v; close Body error: %v", respErr, err)
+			}
+			appendToErrorList(attempt, respErr)
+			//we do not want to process server misconfiguration here
+			break
+		}
+		newLastModified := resp.Header.Get("Last-Modified")
+		if lastModified != "" && newLastModified != lastModified {
+			// last modified changed, retry from the beginning
+			lastModified = newLastModified
+			forceRestart = true
+			continue
+		}
+		if resp.StatusCode == http.StatusOK {
+			// we received StatusOK which is the response for the whole content, not for the partial one
+			rsp.BodyLength = int(resp.ContentLength)
+		}
+		//reset to be not affected by the client.Do timeouts
+		inactivityTimer.Reset(inactivityTimeout)
+		var written int64
+		for {
+			var copyErr error
+
+			written, copyErr = io.CopyN(local, resp.Body, chunkSize)
+			copiedSize += written
+
+			if copyErr != nil {
+				if objSize != copiedSize {
+					if innerCtx.Err() != nil {
+						// the error comes from canceled context, which indicates inactivity timeout
+						appendToErrorList(attempt, fmt.Errorf("inactivity for %s", inactivityTimeout))
+					} else {
+						appendToErrorList(attempt, fmt.Errorf("error from CopyN: %v", copyErr))
+					}
```

Comment:
Wouldn't it make sense to handle io.EOF error differently by printing it as "premature EoF after %d out of %d bytes"? That is more informative then just printing EoF as the error.

Reply:
I added it here: https://github.com/lf-edge/eve/pull/3236/files#diff-438b9bb8c5586aa8dc22614121cdc879f824fb4e290c270f5ff946367d72703fR330

Reply:
Thanks